### ISSUE TITLE
Authentication with OAuth credentials

### ISSUE THREAD

**Issue by JCKodel** [nhost/nhost#2554]
**Is your feature request related to a problem? Please describe.**
When working with apps (Flutter or native), we often use native authentication methods with Google and Apple.

Those methods return an idToken and an accessToken.

Authentication providers such as Firebase Authentication allow us to send those tokens as an OAuth response, thus, authenticating an user and returning the Firebase Authentication JWT.

As far as I could see, nHost only allow OAuth web authentication (this is not a suitable solution for native apps, especially with Apple).

**Describe the solution you'd like**
A method that allow us to send Google and Apple idToken and accessToken for authentication (i.e.: skipping the web oauth authentication).

For Firebase Authentication, this is done as:

```dart
final oAuthProvider =
  OAuthProvider(switch (oauthPrincipal.signInProvider) {
    SignInProvider.apple => AppleAuthProvider.PROVIDER_ID,
    SignInProvider.google => GoogleAuthProvider.PROVIDER_ID,
  });

final credential = oAuthProvider.credential(
  idToken: oauthPrincipal.idToken,
  accessToken: oauthPrincipal.accessToken,
  signInMethod: oAuthProvider.providerId,
);

final auth = await FirebaseAuth.instance.signInWithCredential(credential);

auth.user; // Contains the Firebase Auth user, with firebase user id, name, e-mail and photoURL.
```

```[tasklist]
### Tasks
- [ ] https://github.com/nhost/hasura-auth/pull/576
- [ ] https://github.com/nhost/hasura-auth/pull/577
- [ ] sdk: js
- [ ] sdk: dart
- [ ] update react-native example
- [ ] docs/guide
```

**Comment by dbarrosop** [comment]
Thanks for the feature request. I have just one question:

> this is not a suitable solution for native apps

Do you have any documentation about this? I am asking because this will help prioritize the feature requests (our understanding is that we have users building native apps with flutter for apple without any issues currently)

**Comment by JCKodel** [comment]
> we have users building native apps with flutter for apple without any issues currently

Using native authentication?

I mean, using this???

![image](https://github.com/nhost/nhost/assets/379339/c599d2be-1959-4762-a1df-62dd8040b61b)

This specific image is from https://pub.dev/packages/sign_in_with_apple. It uses the native (by native I mean the Operating System native dialog) auth dialog that only generates an id token and an access token (it is not web, so there is no redirectUrl or anything like that).

For Android, we use https://pub.dev/packages/google_sign_in, that shows the Android native Google Auth dialog:

![image](https://github.com/nhost/nhost/assets/379339/5ae80fe9-aa34-4af0-9be9-39b44dc116bc)

Even the OAuth settings are different in these cases (because there is no web view available in this kind of authentication, so there is no need to redirectUrl. Google, for instance, allows us even to lock the OAuth client to an specific app id).

Both those authentication methods are local and returns an access token for each provider.

Then, with those credentials, we can call Firebase Auth, so it creates an user based on the information on those access tokens (including name, email and photo url).

This is done through this method: https://firebase.google.com/docs/auth/flutter/federated-auth (`signInWithCredential`).

Then, it returns an Firebase User (with an Firebase Id) and we can get an access token from this user (which Hasura documents here: https://hasura.io/blog/authentication-and-authorization-using-hasura-and-firebase/)

As far as I could see, nhost always use an web authentication, which is not appropriate for mobile apps (for instance: we can't create a new account in native authentication, the account management is made by the OS itself (Settings > Accounts in Android)). 

We have an app with +10 million downloads that currently uses Firebase Auth. We have a problem that is importing those users to a relational database, which is a process that costs us a lot of time currently). We are looking for a solution that could just drop in as an replacement for Firebase Auth (meaning: we already do our local authentication using Apple or Google, we just need to get the OAuth tokens generated by these native solutions and create our JWT to access our Hasura - that's where nHost Auth could come in).

TL;DR: We would need nhost to mimic the `signInWithCredentials` from Firebase Auth (https://firebase.google.com/docs/auth/flutter/federated-auth)

**Comment by dbarrosop** [comment]
> Using native authentication?

No, I was asking about your statement "this is not a suitable solution for native apps". We have users building native applications using nhost so I was trying to understand if something had changed that may break applications. From what I am understanding from your last comment you should be able to use nhost already but you would prefer a different authentcation mechanism that leverages some builtin functionality.

**Comment by approachapp** [comment]
Unfortunately, as of the current state, integrating "Sign in with Apple" and "Sign in with Google" directly within mobile applications via Nhost is not supported. The authentication class provided by the Nhost library lacks the necessary methods to facilitate these sign-in processes.

Additionally, implementing a magic link login feature within mobile apps through Nhost requires a reliable method to prompt the email client to open the app, which hinges on App Links and Deep Links for Android, as well as Universal Links and Custom URL Schemes for iOS. 

While App Links and Universal Links offer an easier approach, they are not reliable. 

Deep Links and Custom URL Schemes would be the correct picks, but it is not possible to make it work with the current Nhost setup.

I came to this issue due to problems we are having with magic links and app links not working every time. 
I was trying to look for a workaround to add "Sign in with Apple" and "Sign in with Google" functionalities within native Flutter mobile applications, but is not currently possible.

The approach of using a redirect URL is not the best for mobile apps and makes the authentication process much more challenging.

A simple "Send code to email" and confirm code in the app would make passwordless email much easier for mobile applications.

Regarding google, apple, the example provided by @JCKodel is exactly what we need. We don't want to force the user to do anything outside the app client to log in.

Would be extremely helpful for all mobile devs if you guys could prioritize this issue.

And in the meantime, if there is a current way to sign in with Google and Apple, please show us how in here or update docs or example app from Git Hub.




> > Using native authentication?
> 
> No, I was asking about your statement "this is not a suitable solution for native apps". We have users building native applications using nhost so I was trying to understand if something had changed that may break applications. From what I am understanding from your last comment you should be able to use nhost already but you would prefer a different authentcation mechanism that leverages some builtin functionality, is this correct?

**Comment by marcusrohden** [comment]
> > Using native authentication?
> 
> No, I was asking about your statement "this is not a suitable solution for native apps". We have users building native applications using nhost so I was trying to understand if something had changed that may break applications. From what I am understanding from your last comment you should be able to use nhost already but you would prefer a different authentcation mechanism that leverages some builtin functionality, is this correct?

Unfortunately, as of the current state, integrating "Sign in with Apple" and "Sign in with Google" directly within mobile applications via Nhost is not supported. The authentication class provided by the Nhost library lacks the necessary methods to facilitate these sign-in processes.

Additionally, implementing a magic link login feature within mobile apps through Nhost requires a reliable method to prompt the email client to open the app, which hinges on App Links and Deep Links for Android, as well as Universal Links and Custom URL Schemes for iOS.

While App Links and Universal Links offer an easier approach, they are not reliable.

Deep Links and Custom URL Schemes would be the correct picks, but it is not possible to make it work with the current Nhost setup.

I came to this issue due to problems we are having with magic links and app links not working every time.
I was trying to look for a workaround to add "Sign in with Apple" and "Sign in with Google" functionalities within native Flutter mobile applications, but is not currently possible.

The approach of using a redirect URL is not the best for mobile apps and makes the authentication process much more challenging.

A simple "Send code to email" and confirm code in the app would make passwordless email much easier for mobile applications.

Regarding google, apple, the example provided by @JCKodel is exactly what we need. We don't want to force the user to do anything outside the app client to log in.

Would be extremely helpful for all mobile devs if you guys could prioritize this issue.

And in the meantime, if there is a current way to sign in with Google and Apple, please show us how in here or update docs or example app from Git Hub.

**Comment by dbarrosop** [comment]
> Deep Links and Custom URL Schemes would be the correct picks, but I believe it is not possible to make it work with the current Nhost setup (please correct-me if I'm wrong).

This is an easy fix, please open a new issue.

> A simple "Send code to email" and confirm code in the app would make passwordless email much easier for mobile applications.

You can implement this yourself, just use a custom email template to show the ticket instead of the link, ask the user to enter it on your app and have your app construct the magic link and call it.

**Comment by JCKodel** [comment]
> but you would prefer a different authentcation mechanism

It's not about preference.

It's NOT "different authentication mechanism"

Is the standard default mobile authentication mechanism.

Apps are not websites.
Apps are not websites.
Apps are not websites.

Both Android and iOS always have an account (google or apple). The default standard mobile authentication mechanism just uses those accounts (with some additional popup so the user can confirm it really wants an authentication, of course).

nHost is amazing for web. But, again, apps are not websites.

For apps, we have no other option than Firebase Auth (because Auth0 is expensive as f.... totally out of reality) :(

**Comment by marcusrohden** [comment]
Agreed that is not the best best case but it's still viable. I got signin with Apple working today but had to find a way around it.

Someone commented in another issue on github, and put me on the right direction. Used a web view that would access the URL indicated on the settings at nhost dashboard, then get the redirectionURL with the token and extract from the string.

**Comment by osseonews** [comment]
Just came across this issue and I agree with the original poster, even when this is not a mobile app. In general, as commented the approach of using a redirect URL makes the authentication process much more challenging for all applications. A simple "Send code to email" and confirm code would make everything easier. This is how every single authentication service I currently use works. You get a 6 digit code to your email or phone which you type into a form on the frontend. I posted another issue about this and I really hope this gets implemented, otherwise the auth service is kind of stuck in the "dark" ages with all this clicking of links, redirects etc.

**Comment by rsnider-wisecode** [comment]
Flutter dev here as well. I am moving from Supabase over to nhost due to the built in Hasura integration. It is a weird experience for a user to be authed thru the web for google/apple sign in. I tried working around this nhost limitation by using native aign in and then trying to pass the token back on the callback endpoint, but it wasnt working because of the session stuff built into the auth.

Supabase allows [signing in with a token](https://supabase.com/docs/reference/dart/auth-signinwithidtoken), If nhost had this funcrtionality, then that should enable the use of native auth.

**Comment by dbarrosop** [comment]
@JCKodel we'd appreciate if you stopped with the disparaging and the misinformation. You are not happy with the service, that's fine. But that doesn't entitle you to spread false claims and to make derisive comments. Thanks.

**Comment by rsnider-wisecode** [comment]
Firebase auth is only free up to 50k MAU. Also @JCKodel, talking shit about maintainers is a great way to get your feature requests ignored.

**Comment by JCKodel** [comment]
> Firebase auth is only free up to 50k MAU. Also @JCKodel, talking shit about maintainers is a great way to get your feature requests ignored.

No, it is not.

I have 344k monthly active users and I use Firebase Auth 100% free.

![image](https://github.com/user-attachments/assets/485c76bb-368b-4fc0-9ce9-bd0421984fa2)

**Comment by dbarrosop** [comment]
@rsnider-wisecode the work to support this is planned for September.

By the way, if you want to share more about your move from supabase don't hesitate to post on discussions or join our discord channel. We are always interested in understanding better the reasons behind changing services and what people like/dislike as it helps us understand customers better. Thanks!

**Comment by marcusrohden** [comment]
> @rsnider-wisecode the work to support this is planned for September.
> 

Hi David, any movement on this?
We are currently using a web view in our mobile app to perform Google/Apple sign-in.
It would be great to provide the ability for mobile apps to provide a native experience for the users.
A web view does not provide the best UX.

Thank you

**Comment by dbarrosop** [comment]
Hello,
no news yet. This and a few other tasks around the auth service are still scheduled for September so hopefully we will start working on them soon (we just need to get out of the way some big work we are currently doing in some other parts). Realistically speaking, we may start working on this end of September and have it ready during October, assuming nothing unexpected happens, of course 🤞.

**Comment by dbarrosop** [comment]
Unfortunately this was delayed a bit but we should have an image ready for testing on Monday. If someone can and wants to help testing that'd be very much appreciated.

**Comment by rsnider-wisecode** [comment]
Lemme know how and when!

**Comment by marcusrohden** [comment]
I also would like to test it. please let me know how to proceed.

**Comment by dbarrosop** [comment]
Still working on the docs and a couple of other things but if you want to help out testing this new feature you can currently use the image version:

```
0.35.0-beta003
```

You can see the openapi spec here:

https://github.com/nhost/hasura-auth/pull/576/files#diff-9fe8f7da727ff26c02939aa060e0348d870e6dd95ba433f9f4d53a9a6967aca0

This means you can call `POST /signin/idtoken` passing in the body:

- provider (required): apple|google
- idtoken (required): string
- nonce (optional, unhashed, required if the idtoken includes it): string
- options: same options object for signing up with any other method

If you are using react-native you can check our SDK changes and also our react native example update here:

https://github.com/nhost/nhost/pull/3000

Regarding the nonce, if your library supports it is recommended as it can help against replay attacks. The way it works is as follows:

- you generate some cryptographically unique string (i.e. a uuid or some other random bytes)
- you hash with sha256 the string above and send it to your native authentication library. The library will add the hashed valued to the idtoken
- when calling the auth service you pass the idtoken and the **unhashed** nonce
- the auth service will hash the nonce and compare it with the value in the idtoken

If you have any questions or if something doesn't work as expected let us know. Thanks for your patience here and any help testing the new feature will be highly appreciated. If you are using something other than react-native and can also help providing some snippets of codes as example for our docs let us know as well as that will be even more appreciated than the testing :)

**Comment by rsnider-wisecode** [comment]
Will this be available in flutter?

**Comment by dbarrosop** [comment]
Yes, soon; https://github.com/nhost/nhost-dart/pull/149

**Comment by dbarrosop** [comment]
This is now fully supported: https://docs.nhost.io/guides/auth/sign-in-idtokens
